dspace-uclouvain/src/main/java/org/dspace/uclouvain/rest/BitstreamDownloadRestController.java

```diff
@@ -0,0 +1,196 @@
+                byte[] buffer = new byte[this.bitstreamContentBufferSize];
+
+                InputStream input = bitstreamService.retrieve(context, bitstream);
+                OutputStream output = response.getOutputStream();
+
+                for (int length = 0; (length = input.read(buffer)) > 0;){
+                    output.write(buffer, 0, length);
+                }
+
+                response.flushBuffer();
+                context.restoreAuthSystemState();
+                return new ResponseEntity<>(HttpStatus.OK);
+            } catch (Exception e) {
+                this.logger.error("Could not retrieve bitstream's input stream: " + e.getMessage());
+                return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+            }
+        } else {
+            this.logger.warn("Failed bitstream download attempt; bitstream uuid: " + uuid + "; used hash: " + (hash != null ? hash : "NONE"));
+            return new ResponseEntity<>(HttpStatus.UNAUTHORIZED);
+        }
+    }
+
+    /** 
+     * First retrieve the list of promoters from the item containing the bitstream.
+     * Then hash their email addresses and compare the hash with the one given with the request.
+     * @param ctx: The current DSpace context.
+     * @param bitstream: The bitstream to check the authorization for.
+     * @param hash: The hash to compare with the promoters' hashes.
+     * @return: If one hash match the given one, return true, else return false..
+    */
+    private boolean isAuthorized(Context ctx, Bitstream bitstream, String hash) {
+        try {
+            DSpaceObject parentObject = this.getRootObjectFromBitstream(ctx, bitstream);
+            
+            if (parentObject != null && parentObject instanceof Item) {
+                Item dspaceItem = (Item) parentObject;
+                List<String> hashList = this.getPromotersEmailsAsHash(ctx, dspaceItem);
+
+                if (!hashList.isEmpty() && hashList.contains(hash)){
+                    return true;
+                }
+            }
+            return false;
+        } catch (Exception e) {
+            logger.error("Unhandled exception occurred while checking bitstream access authorization", e);
+            return false;
+        }
+    }
+
+    /**
+     * Retrieve the parent object of a given bitstream.
+     * @param ctx: The current DSpace context.
+     * @param bitstream: The bitstream to retrieve the parent object from.
+     * @return: The parent object of the bitstream if found else return null.
+     */
+    private DSpaceObject getRootObjectFromBitstream(Context ctx, Bitstream bitstream) {
+        try {
+            DSpaceObject dso = this.bitstreamService.getParentObject(ctx, bitstream);
+            return dso;
+        } catch (SQLException e) {
+            logger.error("Could not find the parent object for the bitstream: " + bitstream.getInternalId(), e);
+            return null;
+        }
+    }
```

Comment:
This is not you problem if DSpace loose her mind ;-) remove this method and use `bitstreamService.getParentObject()` into initial method